### ISSUE TITLE
single file publish: AppContext.BaseDirectory doesn't point to apphost directory

### ISSUE THREAD

**Issue by dasMulli** [dotnet/core-setup#7491]
## Steps to reproduce
1. Create a hello world program:
```c#
static void Main(string[] args)
{
    Console.WriteLine("Hello World!");
    Console.WriteLine($"Base Directory: {AppContext.BaseDirectory}");
    Console.WriteLine($"Assembly Location: {typeof(Program).Assembly.Location}");
}
```
2. Publish self contained - `dotnet publish -c Release -r win-x64 -p:PublishSingleFile=True --self-contained false` (or self-contained)

3. Run the resulting exe file.

## Expected  behavior

According to the [design document](https://github.com/dotnet/designs/blob/686205657ace53b3e6d568cb098be302f6974956/accepted/single-file/design.md), `AppContext.BaseDirectory` is the API that developers should use to access the path where the app host resides.

The app context base directory should therefore not be in `AppData\Local\Temp\.net\`.

Background: I want to use this directory to load override config files.

## Actual behavior

```
C:\demos\testcons\bin\Release\netcoreapp3.0\win-x64\publish>testcons.exe
Hello World!
Base Directory: C:\Users\ullrimar\AppData\Local\Temp\.net\testcons\53kpwgv0.3mq\
Assembly Location: C:\Users\ullrimar\AppData\Local\Temp\.net\testcons\53kpwgv0.3mq\testcons.dll
```

## Environment data

```
>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview7-012821
 Commit:    6348f1068a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview7-012821\

Host (useful for support):
  Version: 3.0.0-preview7-27912-14
  Commit:  4da6ee6450
```

cc @vitek-karas

**Comment by dasMulli** [comment]
@swaroop-sridhar is there any other path at the moment that points to the app host directory and/or name?

**Comment by swaroop-sridhar** [comment]
In the current version, we simply extract the contents of the bundle to a directory on the disk, and declare that directory as the base-directory. This choice aims to preserve compatibility for apps that assume that apcontext.basedirectory contains the app.dll. In the current version, we'll need to use native APIs such as [GetModuleFileName](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea) to get the path of the native AppHost.

In subsequent versions, once the app.dll is processed directly from the bundle, the design doc discusses options to set `AppContext.BaseDirectory`. The current proposal is to set it to the location of the AppHost.

**Comment by swaroop-sridhar** [comment]
You could also try using: `Process.GetCurrentProcess().MainModule.FileName` as suggested here: https://github.com/dotnet/coreclr/issues/20287#issuecomment-506590901

**Comment by dasMulli** [comment]
That works, thanks!

Is this (esp. `AppContext.BaseDirectory`) subject to change? If not, I'd ask for the design document to be in line with the actual implementation, or hint at which parts are part of .NET Core 3.0.

**Comment by swaroop-sridhar** [comment]
Yes @dasMulli the value of `AppContext.BaseDirectory` is open question in the next version of .net core, -- because when we load assemblies directly from the bundle, there's no extraction directory.

**Comment by karataliu** [comment]
Can make `AppContext.BaseDirectory` an independent way for getting run root?
If using `Process.GetCurrentProcess().MainModule.FileName`, it works fine with single file app, but when run in vs test, the value returned is 'dotnet.exe'.
